### PR TITLE
[FS-926] Create new conversation type for global team conversation

### PR DESCRIPTION
**Pull request by elland**
This should cover [FS-927](https://wearezeta.atlassian.net/browse/FS-927), as the endpoint has been added.

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: services/galley/src/Galley/API/Teams.hs
Comment (elland) [inline services/galley/src/Galley/API/Teams.hs:246]:
```suggestion
```

File: services/galley/src/Galley/Cassandra/Access.hs
Comment (elland) [inline services/galley/src/Galley/Cassandra/Access.hs:34]: Double check this with team.

File: services/galley/src/Galley/API/Internal.hs
Comment (elland) [inline services/galley/src/Galley/API/Internal.hs:688]: Double check this with team.

File: libs/wire-api/src/Wire/API/Conversation.hs
Comment (elland) [inline libs/wire-api/src/Wire/API/Conversation.hs:515]: @pcapriotti can you confirm this is the right access role?

File: services/galley/src/Galley/Cassandra/Conversation.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/Cassandra/Conversation.hs:209]: I think this should be `cnvmAccess = [InviteAccess]`, as users should be able to join only by someone else adding them (namely, the team admin). `LinkAccess` means it is sufficient to know the conversation ID.

File: services/galley/src/Galley/Cassandra/Conversation.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/Cassandra/Conversation.hs:210]:
```suggestion
                cnvmAccessRoles = teamAccessRole,
```
The conversation should be accessible only to team members.

File: services/galley/src/Galley/Cassandra/Conversation.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/Cassandra/Conversation.hs:211]:
```suggestion
                cnvmName = Just "Global team conversation",
```

File: services/galley/src/Galley/Cassandra/Access.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/Cassandra/Access.hs:34]:
```suggestion
defAccess GlobalTeamConv s = maybe [InviteAccess] Set.toList s
```
This seems to be in line with the last catch-all clause for a set with at least one element. If the input set is `Nothing`, team members should be able to join via an invitation.

File: services/galley/test/integration/API/MLS.hs
Comment (mdimjasevic) [inline services/galley/test/integration/API/MLS.hs:2072]: I don't understand what queues we're dealing with here. Is that necessary?

File: services/galley/test/integration/API/MLS.hs
Comment (mdimjasevic) [inline services/galley/test/integration/API/MLS.hs:2083]: ditto

**Review by mdimjasevic — COMMENTED** [review]
Looks good! I'd ask you to change a couple of things as I suggested inlined. Once you address that I think this is good to merge.

Also, I didn't see a point where you need a creator client ID. I'm puzzled. This should have shown up somewhere.

File: services/galley/src/Galley/Cassandra/Conversation.hs
Comment (elland) [reply services/galley/src/Galley/Cassandra/Conversation.hs:209]: Yeah, I thought you would join by `convId`, by querying the `GET` endpoint for global team convs. This is sadly not specified in the spec either.

File: libs/wire-api/src/Wire/API/Conversation.hs
Comment (mdimjasevic) [inline libs/wire-api/src/Wire/API/Conversation.hs:222]: Now that I think about it, if we're to stay consistent with how we've implemented MLS so far, this should go somewhere under `Wire.API.MLS`, perhaps in a new module.

File: services/galley/src/Galley/API/Teams.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/API/Teams.hs:257]: After creating the global team conversation, you have to add the creator client `c` to the group with something like this:

```haskell
  E.addMLSClients lcnv (qUntagged lusr) (Set.singleton (c, nullKeyPackageRef))
```
The same goes for the binding team.

File: services/galley/src/Galley/Cassandra/Conversation.hs
Comment (mdimjasevic) [reply services/galley/src/Galley/Cassandra/Conversation.hs:209]: Then the list is `[InviteAccess, LinkAccess]`, right?

File: services/galley/src/Galley/Cassandra/Access.hs
Comment (elland) [reply services/galley/src/Galley/Cassandra/Access.hs:34]: But you shouldn't be invited to the global team conv, right? It should be done automatically when you join a team?

**Review by mdimjasevic — DISMISSED** [review]
I can't see any changes to the Message module in Galley, hence it can't be this PR is ready. There in commit processing you'd need to add the creator client in the first epoch. Paolo has been refactoring that module and introducing a change like this, which I need too in the self-conversation PR.

File: libs/wire-api/src/Wire/API/MLS/GlobalTeamConversation.hs
Comment (pcapriotti) [inline libs/wire-api/src/Wire/API/MLS/GlobalTeamConversation.hs:50]: Better to use the same label and format for the ID as for normal conversations.

File: services/galley/src/Galley/API/Federation.hs
Comment (elland) [inline services/galley/src/Galley/API/Federation.hs:282]: For now it might make more sense to throw a 403 instead of an empty list.

File: services/galley/src/Galley/API/Internal.hs
Comment (elland) [inline services/galley/src/Galley/API/Internal.hs:717]: Actually we need a different logic here. For the global team conv we need to do the MLS client removals regardless of membership, since that's implicit.

File: services/galley/src/Galley/API/Query.hs
Comment (elland) [inline services/galley/src/Galley/API/Query.hs:182]: Don't actually create the MLS clients here. Instead add logic in the commit handling code to deal with the global team conversation special case. (Wait for refactoring to be merged, #2844 )

File: services/galley/src/Galley/API/Teams.hs
Comment (elland) [inline services/galley/src/Galley/API/Teams.hs:261]: Forgot to clear this after debug. Weird that hlint didn't complain.

File: services/galley/src/Galley/Cassandra/Conversation.hs
Comment (elland) [inline services/galley/src/Galley/Cassandra/Conversation.hs:197]: Rewrite this to not actually query the member list.

File: services/galley/src/Galley/Cassandra/Conversation.hs
Comment (elland) [inline services/galley/src/Galley/Cassandra/Conversation.hs:219]: If we use GlobalTeamConversation directly we can simplify this.

File: services/galley/test/integration/API.hs
Comment (elland) [inline services/galley/test/integration/API.hs:3840]: Missed reverting this line from ormolu mess.

File: services/galley/test/integration/API/MLS.hs
Comment (elland) [inline services/galley/test/integration/API/MLS.hs:199]: Add test for self-invite, MLS stuff.

**Review by pcapriotti — CHANGES_REQUESTED** [review]
There needs to be some special logic either when listing conversations (to exclude the global team and self conversations) or when returning them (to convert them to the correct JSON format).

File: services/galley/src/Galley/API/Federation.hs
Comment (elland) [reply services/galley/src/Galley/API/Federation.hs:282]: Actually this is federation, we don't throw errors.

File: services/galley/src/Galley/API/Action.hs
Comment (pcapriotti) [inline services/galley/src/Galley/API/Action.hs:286]: Why the extra `do`?

File: services/galley/src/Galley/API/Action.hs
Comment (pcapriotti) [inline services/galley/src/Galley/API/Action.hs:673]: (*Referring to this whole function*)

This is not so terrible, but the GTC querying bits could be moved to a different function to cleanly separate the special logic for GTCs from the rest of the logic.

Of course, the "clean" haskelly solution would be to make `Conversation` a sum type encompassing the two kinds of conversations (i.e. regular or GTC). Given the unfriendliness of record types in Haskell, I understand you might have some reservations about going this route, but maybe it's worth considering.

File: services/galley/src/Galley/API/MLS/Message.hs
Comment (pcapriotti) [inline services/galley/src/Galley/API/MLS/Message.hs]: The `undefined` here is not good, just a horrible runtime bug waiting to happen. Also, this function could be extracted and reused, since it (or a version of it) is defined at least twice. 

If you decide not to go with the sum type solution for the `Conversation` type, then I think the only way would be to make the creator field optional, and adapt the code everywhere else.

File: services/galley/src/Galley/API/Query.hs
Comment (pcapriotti) [inline services/galley/src/Galley/API/Query.hs]: You're still setting the client on on GET, which is prone to race conditions. The `nullKeyPackageRef` hack should not be used anymore.

File: services/galley/src/Galley/Cassandra/Conversation.hs
Comment (pcapriotti) [inline services/galley/src/Galley/Cassandra/Conversation.hs:255]: Why the duplication? Can't `getGlobalTeamConversation` simply call `getGlobalTeamConversationById`?

**Review by pcapriotti — CHANGES_REQUESTED** [review]
There are a few issues, see comments below.

I can't find the code that produces the correct rendering of GTCs when fetched through the normal `POST /conversations/list` API. There is also no test for listing.

And one more thing that seems to be missing is creating the GTC entry on the `list-ids` endpoint, so that clients don't need to rely on the mutating semantics of the GET endpoint, just like what we discussed for the self conversation.

File: services/galley/src/Galley/API/Action.hs
Comment (elland) [reply services/galley/src/Galley/API/Action.hs:286]: Probably a remnant!

File: services/galley/src/Galley/API/Query.hs
Comment (elland) [reply services/galley/src/Galley/API/Query.hs]: I think this was a rebase mix-up. Good catch.

Comment (elland) [reply services/galley/src/Galley/API/Query.hs]: Oh, more like I missed a spot.

File: services/galley/src/Galley/API/Action.hs
Comment (elland) [reply services/galley/src/Galley/API/Action.hs]: > make Conversation a sum type 

I'm all for it but this PR is large enough as it is. Should we create a follow up?

File: services/galley/src/Galley/API/MLS/Message.hs
Comment (pcapriotti) [inline services/galley/src/Galley/API/MLS/Message.hs:487]: Wouldn't it be better to get the conversation first, then check if it's the GTC looking at its type? That way the normal case of conversations doesn't have to go through yet another db hit.

File: services/galley/src/Galley/API/MLS/Message.hs
Comment (pcapriotti) [inline services/galley/src/Galley/API/MLS/Message.hs:484]: Won't this return the empty list?

File: services/galley/src/Galley/API/MLS/Util.hs
Comment (pcapriotti) [inline services/galley/src/Galley/API/MLS/Util.hs:80]: This is a bit strange. Can't we simply accept that a GTC does not have a creator? All this fudging around with the creator should not be necessary.

**Review by pcapriotti — APPROVED** [review]
Looks generally good, but I would consider changing the way the creator of a GTC is handled. Since a GTC doesn't really have a creator, I would just try to make the code reflect that, instead of forcing one. That means changing the types around a bit, but it shouldn't be a huge effort.

I'll approve for now and let you decide if it's worth doing that or not.

File: services/galley/src/Galley/API/MLS/Message.hs
Comment (elland) [reply services/galley/src/Galley/API/MLS/Message.hs:487]: Actually this code is now outdated, since I've changed how getting members work further down the chain, we can basically revert this.

File: services/galley/src/Galley/API/MLS/Message.hs
Comment (elland) [reply services/galley/src/Galley/API/MLS/Message.hs:484]: It now returns the same results as they member list already in the conversation. Dead code.

File: services/galley/src/Galley/API/MLS/Util.hs
Comment (elland) [reply services/galley/src/Galley/API/MLS/Util.hs:80]: This has large ramifications… I think it makes sense to clean it up when we do the conversation sum type.

File: services/galley/test/integration/API/Util.hs
Comment (jschaul) [inline services/galley/test/integration/API/Util.hs:291]: why all this reformatting as part of this PR? Makes it much harder to see actual changes that happened.

File: services/galley/src/Galley/API/MLS/Util.hs
Comment (jschaul) [inline services/galley/src/Galley/API/MLS/Util.hs:80]: I don't like the sideeffects introduced here into each lookup of a conversation. This looks badly designed.

Elsewhere, it seems global team conversation id is the same as the team Id (even though this is not documented anywhere; and it means there can only ever be one single global team conversation). Can't this mean you can tell earlier, if you have a user object available, you'd know their team, and already know which kind of conv you're dealing with, without an extra lookup?

And why does a GET of any conversation ID lead to an insert of a random user as creator of this conversation? That makes no sense semantically.

Comment (elland) [reply services/galley/src/Galley/API/MLS/Util.hs:80]: We're taking this out in a follow-up  👍

File: services/galley/test/integration/API/Util.hs
Comment (elland) [reply services/galley/test/integration/API/Util.hs:291]: We had some issues with ormolu, I might have missed some places.

**Pull request merged**
